### PR TITLE
Update freezebalance

### PR DESCRIPTION
**Pull request by serderovsh**
"freezebalance", "unfreezebalance" did not work

### CODE REVIEW

**Comment by serderovsh** [comment]
@Kondax, Duration check?
I added an exception in case of errors
```
if(transaction.Error) {
     return callback(transaction.Error);
}
```
I will think about the structure for the future. "TransactionBuilder", identical conditions are repeated in both trx.js and transactionBuilder.js.

**Comment by Kondax** [comment]
To be honest error checking isn't necessary in trx.js as it's also done in transactionBuilder.js as you mentioned.

File: src/lib/transactionBuilder.js [hunk 1]
```diff
@@ -122,19 +122,18 @@ export default class TransactionBuilder {
         }).catch(err => callback(err));
     }
 
-    freezeBalance(address = this.tronWeb.defaultAddress.hex, amount = 0, duration = 3, resource = "BANDWIDTH", callback = false) {
-        if(utils.isFunction(duration)) {
-            callback = duration;
-            duration = 3;
+    freezeBalance(amount = 0, duration = 3, resource = "BANDWIDTH", address = this.tronWeb.defaultAddress.hex, callback = false)
```
Comment (Kondax) [inline src/lib/transactionBuilder.js]: Both `duration` and `resource` can be functions. You should check if they are functions, and if they are set `callback = func`, and set their default values.

File: src/lib/transactionBuilder.js [hunk 2]
```diff
@@ -158,23 +157,21 @@ export default class TransactionBuilder {
         }).catch(err => callback(err));
     }
 
-    unfreezeBalance(address = this.tronWeb.defaultAddress.hex, resource = "BANDWIDTH", callback = false) {
+    unfreezeBalance(resource = "BANDWIDTH", address = this.tronWeb.defaultAddress.hex, callback = false) {
```
Comment (Kondax) [inline src/lib/transactionBuilder.js]: Same needs to be done here.

File: src/lib/trx.js [hunk 3]
```diff
@@ -675,6 +675,106 @@ export default class Trx {
...
+     * Will give bandwidth OR Energy and TRON Power(voting rights)
+     * to the owner of the frozen tokens.
+     *
+     * @param amount - is the number of frozen trx
+     * @param duration - is the duration in days to be frozen
+     * @param resource - is the type, must be either "ENERGY" or "BANDWIDTH"
+     * @param options
+     * @param callback
+     */
+    async freezeBalance(amount = 0, duration = 3, resource = "BANDWIDTH", options = {}, callback = false)
```
Comment (Kondax) [inline src/lib/trx.js]: Here too.

File: src/lib/trx.js [hunk 4]
```diff
@@ -675,6 +675,106 @@ export default class Trx {
...
+
+    /**
+     * Unfreeze TRX that has passed the minimum freeze duration.
+     * Unfreezing will remove bandwidth and TRON Power.
+     *
+     * @param resource - is the type, must be either "ENERGY" or "BANDWIDTH"
+     * @param options
+     * @param callback
+     */
+    async unfreezeBalance(resource = "BANDWIDTH", options = {}, callback = false)
```
Comment (Kondax) [inline src/lib/trx.js]: Also here.

**Comment by serderovsh** [comment]
> To be honest error checking isn't necessary in trx.js as it's also done in transactionBuilder.js as you mentioned.

I originally wrote the condition in both places, and it looked weird.
Now I understand everything

**Comment by serderovsh** [comment]
I will do the settings in the "transactionBuilder" and in the "trx.js" I will transmit the finished results

File: src/lib/transactionBuilder.js (hunk 1 above)
Comment (serderovsh) [reply src/lib/transactionBuilder.js]: Made

File: src/lib/trx.js (hunk 3 above)
Comment (serderovsh) [reply src/lib/trx.js]: Made

File: src/lib/transactionBuilder.js (hunk 1 above)
Comment (serderovsh) [reply src/lib/transactionBuilder.js]: The basic checks I have are transactionBuilder, and trx.js accepts the results and signs the transaction.

> I already added all checks.

**Comment by kookiekrak** [comment]
lets avoid breaking changes to existing api if possible. we can do a v3 branch and fix stuff there

**Comment by serderovsh** [comment]
Okay

**Pull request closed without merging**